File: Data Structures & Algorithms/decode-ways/submission-11.java

class Solution {
    int res = 0;
    Integer[] dp;
    public int numDecodings(String s) {
        dp = new Integer[s.length()];
        return dfs(0,s);
        
        
    }

    public int dfs(int i,String s){
        if(i==s.length()){
            return 1;
        }
        if(s.charAt(i)=='0'){
            return 0;
        }
        if(dp[i]!=null){
            return dp[i];
        }
        int value =i+1<s.length()?Integer.parseInt(s.substring(i,i+2)):0;
        if(value>0 && value<=26){
            dp[i] = dfs(i+1,s)+dfs(i+2,s);
        } else {
            dp[i] = dfs(i+1,s);
        }
        return dp[i];
    }

}

File: Data Structures & Algorithms/count-subsequences/submission-11.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        N,M = len(s),len(t)
        dp = {}
        def dfs(i,j):
            if j == M:
                return 1
            if i == N:
                return 0
            if (i,j) in dp:
                return dp[(i,j)]
            dp[(i,j)] = dfs(i+1,j)
            if s[i] == t[j]:
                dp[(i,j)] += dfs(i+1,j+1)
            return dp[(i,j)]
        return dfs(0,0)


        
        # caaat
        # i
        # cat
        # j
        # if i== j
        # i+1,j+1 + i+1,j
        # if i!=j
        # i+1
        # j
        